drivers/video/video_pio_dma.c

```diff
@@ -0,0 +1,549 @@
+/*
+ * Copyright The Zephyr Project Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT raspberrypi_video_pio_dma
+
+#include <hardware/pio.h>
+#include <zephyr/drivers/dma.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/misc/pio_rpi_pico/pio_rpi_pico.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/video-controls.h>
+#include <zephyr/drivers/video.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/util.h>
+
+#include "video_ctrls.h"
+#include "video_device.h"
+
+LOG_MODULE_REGISTER(video_pio_dma, CONFIG_VIDEO_LOG_LEVEL);
+
+#define VIDEO_PIO_DMA_STACK_SIZE 1024
+#define VIDEO_PIO_DMA_PRIORITY   5
+#define MAX_FRAME_RATE           10
+#define MIN_FRAME_RATE           1
+#define VIDEO_PIO_DMA_STACK_SIZE 1024
+#define VIDEO_PIO_DMA_PRIORITY   5
+#define VIDEO_PIO_DMA_RX_BIT     0
+#define VIDEO_PIO_DMA_RELOAD_BIT 1
+#define VIDEO_PIO_DMA_ERR_BIT    2
+
+K_THREAD_STACK_DEFINE(video_pio_dma_stack_area, VIDEO_PIO_DMA_STACK_SIZE);
+
+struct video_pio_dma_data {
+	const struct device *dev;
+	PIO pio;
+	size_t pio_sm;
+	int pio_program_offset;
+	pio_sm_config pio_sm_config;
+	pio_program_t pio_program;
+	struct k_fifo fifo_in;
+	struct k_fifo fifo_out;
+	struct k_work_delayable work;
+	struct k_work_q work_q;
+	struct k_poll_signal *sig;
+	struct video_buffer *vbuf;
+	int dma_channel;
+	unsigned int frame_rate;
+	unsigned int pclk_per_px;
+	unsigned int border_px;
+	struct video_format fmt;
+	atomic_t stream_evt;
+	bool stream_on;
+};
+
+struct video_pio_dma_config {
+	const struct device *pio_dev;
+	const struct pinctrl_dev_config *pin_cfg;
+	const struct gpio_dt_spec vsync_gpio;
+	const struct gpio_dt_spec hsync_gpio;
+	const struct gpio_dt_spec pclk_gpio;
+	const struct gpio_dt_spec data_gpio;
+	const struct device *sensor_dev;
+	const struct device *dma_dev;
+	const uint32_t data_bits;
+};
+
+#ifdef CONFIG_POLL
+static int video_pio_dma_set_signal(const struct device *dev, struct k_poll_signal *sig)
+{
+	struct video_pio_dma_data *data = dev->data;
+
+	if (data->sig && sig != NULL) {
+		return -EALREADY;
+	}
+	data->sig = sig;
+	return 0;
+}
+#endif
+
+static int video_pio_dma_config_pio(const struct device *dev)
+{
+	const struct video_pio_dma_config *config = dev->config;
+	struct video_pio_dma_data *data = dev->data;
+
+	if (data->pio_program_offset >= 0) {
+		pio_remove_program(data->pio, &data->pio_program, data->pio_program_offset);
+		data->pio_program_offset = -1;
+	}
+	uint16_t pio_program_instructions[] = {
+		/* istruction 0 */
+		pio_encode_pull(false, true),
+		/* istruction 1 */
+		pio_encode_wait_gpio(false, config->vsync_gpio.pin),
+		/* istruction 2 */
+		pio_encode_wait_gpio(true, config->vsync_gpio.pin),
+		/* istruction 3 */
+		pio_encode_set(pio_y, data->border_px - 1),
```

Comment:
Given `data->border_px` can be any variable, this gives us the opportunity to select it at runtime to allow arbitrary cropping by this driver.

This allows to use [Linux-style selection API](https://docs.kernel.org/userspace-api/media/v4l/selection-api.html) on the RP2040 receiver!

This is just a side-note, better keep the PR simple and anyone is free to iterate on top of it.